[runtime] Add support for hierarchical tasks

andresilva opened the pull request:
This PR adds support for hierarchical task management where child tasks are automatically aborted when their parent task completes or is aborted. The implementation introduces a `spawn_child` method to the `Spawner` trait, which creates parent-child relationships between tasks using `futures::future::AbortHandle` for cleanup.

A key principle is that parent-child relationships are tied to the task that spawned them, not to the context itself. When you clone a context (via `Clone::clone` or methods like `with_label`), you get another reference to the same context. This means:
- Tasks spawned with `spawn` from any context (original or cloned) are always independent
- Only tasks spawned with `spawn_child` become children of the current task
- Child tasks are tied to the task that spawned them, not to the context

This design ensures clean separation between different task scopes while maintaining the parent-child abort semantics within each scope. The `spawn_child` method takes `&self` rather than consuming the context, allowing multiple children to be spawned from the same parent context. Unlike `spawn` and `spawn_ref`, `spawn_child` does not set `spawned` to true, as the parent context can legitimately spawn multiple children.

This PR does not migrate any of the existing subsystems to use child tasks. I think this should be done in a follow-up PR (and would require some guidance).

Fixes #1100.

patrick-ogrady commented:
This is pretty neat. Haven't had the time to review it yet but thanks for taking a stab.

BrendanChou commented:
> A key principle is that a `Context` represents a logical scope rather than just a task execution environment - when you clone a context, the clone shares the same scope and children list as the original, meaning all child tasks spawned from any clone of the same context are tracked together and will be aborted as a group.

Forgive me if I don't understand correctly, but then can't there be two contexts that spawn two separate tasks but share `children`? Meaning if one of the two tasks aborts, then ALL children from BOTH tasks are aborted?

andresilva commented:
@BrendanChou I have updated the PR, I believe it now has cleaner (and less surprising) semantics. I also updated the PR description to reflect the changes.

andresilva commented on runtime/src/deterministic.rs:
```diff
@@ -738,22 +791,42 @@ impl crate::Spawner for Context {
+        // Create child context with its own empty children list
+        let child_context = Self {
+            name: self.name.clone(),
+            spawned: self.spawned,
```
The correct value is to set spawned to false since this is the new context for the child (which hasn't spawned anything yet since it's just being created). Both `deterministic` and `tokio` set this to false explicitly now, although in practice `self.spawned` must have been `false`, so either case is correct.

andresilva commented:
This was a bit tricky but I think the semantics are now correct (and all cases are covered by tests I think):

```rust
context.spawn(|context| { // task 1
  context.spawn_child({ ... }) // child 1
  context.spawn({ ... }) // task 2
})
```

Child 1 and Task 2 are independent. If Task 1 dies then Child 1 does as well, while Task 2 keeps going. If Task 2 dies then Child 1 and Task 1 are unaffected.

```rust
context.spawn(|context| { // task 1
  context.spawn_ref()({ ... }) // task 2
  context.spawn_child({ ... }) // child 1
})
```

Same as above.

```rust
context.spawn(|context| { // task 1
  context.spawn_child({ ... }) // child 1
  context.clone().spawn_child({ ... }) // child 2
})
```
If Task 1 dies then both Child 1 and 2 die as well. Cloned contexts still inherit the same children list.

```rust
context.clone().spawn(|context| { // task 1
  context.spawn_child({ ... }) // child 1
})

context.spawn(|context| { ... }) // task 2
```

If Task 2 dies then Child 1 is unaffected.

patrick-ogrady commented:
> This was a bit tricky but I think the semantics are now correct (and all cases are covered by tests I think):
> 
> ```rust
> context.spawn(|context| { // task 1
>   context.spawn_child({ ... }) // child 1
>   context.spawn({ ... }) // task 2
> })
> ```
> 
> Child 1 and Task 2 are independent. If Task 1 dies then Child 1 does as well, while Task 2 keeps going. If Task 2 dies then Child 1 and Task 1 are unaffected.
> 
> ```rust
> context.spawn(|context| { // task 1
>   context.spawn_ref()({ ... }) // task 2
>   context.spawn_child({ ... }) // child 1
> })
> ```
> 
> Same as above.
> 
> ```rust
> context.spawn(|context| { // task 1
>   context.spawn_child({ ... }) // child 1
>   context.clone().spawn_child({ ... }) // child 2
> })
> ```
> 
> If Task 1 dies then both Child 1 and 2 die as well. Cloning the context makes no difference since the inheritance is at the task level.
> 
> ```rust
> context.clone().spawn(|context| { // task 1
>   context.spawn_child({ ... }) // child 1
> })
> 
> context.spawn(|context| { ... }) // task 2
> ```
> 
> If Task 2 dies then Child 1 is unaffected.

This looks correct to me.

patrick-ogrady commented on runtime/src/deterministic.rs:
```diff
@@ -747,15 +757,38 @@ impl crate::Spawner for Context {
+        T: Send + 'static,
+    {
+        // Create child context
+        let child_context = self.clone();
```
I actually prefer requiring `spawn_child` consume `self` here (like a normal context spawn).

It should be up to the user to ensure they clone the context for any child (this also allows them to provide a `with_label` here.

patrick-ogrady commented on runtime/src/tokio/runtime.rs:
```diff
@@ -418,15 +427,42 @@ impl crate::Spawner for Context {
+        T: Send + 'static,
+    {
+        // Create child context
+        let child_context = self.clone();
```
See note in `deterministic` about `self.clone()`.

patrick-ogrady commented on runtime/src/deterministic.rs:
```diff
@@ -747,15 +757,38 @@ impl crate::Spawner for Context {
+        T: Send + 'static,
+    {
+        // Store parent's children list
+        let parent_children = self.children.clone();
```
While this may look like a lot of cloning, this is just behind an `Arc` (and its more efficient than some alternatives that avoid this call).
